Fix export to geopdf skips raster and mesh layers

nyalldawson opened the pull request:
Fixes #32148

nyalldawson committed:
Fix export to geopdf skips raster and mesh layers

Fixes #32148

nirvn commented:
@nyalldawson , looking good. Is there no additional test case here due to travis not supporting GeoPDF export yet?

nyalldawson committed:
Add unit tests for raster rendering during staged render jobs

nyalldawson commented:
@nirvn I've just added one

nyalldawson committed:
Test mask

PeterPetrik commented on src/core/qgsmaplayerrenderer.h:
```diff
@@ -66,6 +67,13 @@ class CORE_EXPORT QgsMapLayerRenderer
     QString layerId() const { return mLayerID; }
 
+    /**
+     * Resets the painter to use for the job to a new \a painter.
```
how it is with ownership of the painter?

question: don't we kind of assume the painter cannot be null in the renderers?

thinking loud: I am wondering if the rendering context should not be part of the `QgsMapLayerRenderer`, since now we have a virtual setter in base class that does not in principle contains at all the context/painter.

nyalldawson replied:
>  I am wondering if the rendering context should not be part of the QgsMapLayerRenderer

it probably should be, yes...

nyalldawson replied:
@PeterPetrik does c1d156d look better to you? I've moved the render context member to the QgsMapLayerRenderer base class. There's a further cleanup which should be done in QGIS 4, which is to make the base class member a reference rather than a pointer, but I can't do that without breaking stable API.

PeterPetrik replied:
looks right, thanks

nyalldawson committed:
Cleanup QgsMapLayerRenderer api

Make associated QgsRenderContext a base class member, since all
renderer subclasses require this
